src/main/java/com/wavefront/sdk/direct/ingestion/WavefrontDirectIngestionClient.java

```diff
@@ -405,4 +415,54 @@ public synchronized void close() {
       logger.log(Level.WARNING, "shutdown error", ex);
     }
   }
+
+  /**
+   * Dequeue and return a batch of at most N items from buffer (where N = batchSize), broken into
+   * chunks where each chunk has at most M bytes of data (where M = messageSizeBytes).
+   *
+   * Visible for testing.
+   *
+   * @param buffer            The buffer queue to retrieve items from.
+   * @param batchSize         The maximum number of items to retrieve from the buffer.
+   * @param messageSizeBytes  The maximum number of bytes in each chunk.
+   * @param dropped           A counter counting the number of items that are dropped.
+   * @return A batch of items retrieved from buffer.
+   */
+  static List<List<String>> getBatch(LinkedBlockingQueue<String> buffer, int batchSize,
+                                       int messageSizeBytes, WavefrontSdkCounter dropped) {
+    batchSize = Math.min(buffer.size(), batchSize);
+    List<List<String>> batch = new ArrayList<>();
+    List<String> chunk = new ArrayList<>();
+    int numBytesInChunk = 0;
+    int count = 0;
+
+    while (count < batchSize) {
+      String item = buffer.poll();
+      if (item == null) {
```

Comment:
why would item be null when you already have `batchSize` protection? Is buffer a shared Object with a different thread?

Reply:
it's only a single thread, but i figured it be good form to account for null regardless.